lib/puma/cluster.rb

```diff
@@ -578,27 +636,25 @@ def wait_workers
 
       @workers.reject! do |w|
         next false if w.pid.nil?
-        begin
-          # We may need to check the PID individually because:
-          # 1. From Ruby versions 2.6 to 3.2, `Process.detach` can prevent or delay
-          #    `Process.wait2(-1)` from detecting a terminated process: https://bugs.ruby-lang.org/issues/19837.
-          # 2. When `fork_worker` is enabled, some worker may not be direct children,
-          #    but grand children.  Because of this they won't be reaped by `Process.wait2(-1)`.
-          if reaped_children.delete(w.pid) || Process.wait(w.pid, Process::WNOHANG)
-            true
-          else
-            w.term if w.term?
-            nil
-          end
-        rescue Errno::ECHILD
-          begin
-            Process.kill(0, w.pid)
-            # child still alive but has another parent (e.g., using fork_worker)
-            w.term if w.term?
-            false
-          rescue Errno::ESRCH, Errno::EPERM
-            true # child is already terminated
-          end
+        # We may need to check the PID individually because:
+        # 1. From Ruby versions 2.6 to 3.2, `Process.detach` can prevent or delay
+        #    `Process.wait2(-1)` from detecting a terminated process: https://bugs.ruby-lang.org/issues/19837.
+        # 2. When `fork_worker` is enabled, some worker may not be direct children,
+        #    but grand children.  Because of this they won't be reaped by `Process.wait2(-1)`.
+        if reaped_children.delete(w.pid) || check_process_terminated(w.pid)
```

Comment:
moved the complex begin/rescue logic to `check_process_terminated`, it's not _exactly_ equivalent but functionally should be (none of the different code can raise the exceptions that would be rescued)